Binary Search Module Support (#833)
cpinitiative/usaco-guide: content/3_Silver/Binary_Search.mdx

</CPPSection>

<JavaSection>

```java
import java.util.function.Predicate;

public class BinarySearch {
	public static void main(String[] args) {
		System.out.println(lastTrue(2,10,(x) -> true)); // 10
		// all numbers satisfy the condition
		System.out.println(lastTrue(2,10,(x) -> x*x<=30)); // 5
		System.out.println(lastTrue(2,10,(x) -> false)); // 1
		// no numbers satisfy the condition
	}

	public static int lastTrue(int lo, int hi, Predicate<Integer> f) {
		for (--lo; lo < hi; ) {
			int mid = lo+(hi-lo+1)/2;
			// find the middle of the current range (rounding up)
			if (f.test(mid)) lo = mid;
				// if mid works, then all numbers smaller than mid also work
			else hi = mid-1;
			// if mid does not work, greater values would not work too
			// so we don't care about them

		}
		return lo;
	}
}
```

> See [Java Function Interface](https://docs.oracle.com/javase/8/docs/api/java/util/function/Predicate.html) if you're not familiar with the Predicate interface used.

Reviewer: ```suggestion
See [Java Function Interface](https://docs.oracle.com/javase/8/docs/api/java/util/function/Predicate.html) if you're not familiar with the `Predicate` interface used.
```

		mid = (lo+hi)//2 # find the middle of the current range
		if f(mid):
			# if mid works, then all numbers smaller than mid also work
			res = mid # update the answer
			lo = mid+1 # cut the part before mid
			# notice that we already handled mid itself so we cut it as well
		else:
			# if mid does not work, greater values would not work too
			hi = mid-1 # cut mid and after
	# now res is the answer
	# if res is l-1 that means no value in the range satisfies the condition
	return res
print(lastTrue(2, 10,lambda x:True)) # 10
# all numbers satisfy the condition
print(lastTrue(2, 10,lambda x:x*x <= 30)) # 5

> print(lastTrue(2, 10,lambda x:False)) # 1

Reviewer: Add a space after `10,` (also please fix the spacing in the lambdas if you don't mind)

# no numbers satisfy the condition
```
See [Lambda Expressions](https://www.w3schools.com/python/python_lambda.asp) if you're not familiar with the syntax used in the program.

</PySection>

</LanguageSection>

#### Implementation 2

> This approach is based on interval jumping. Essentially, we start from the beginning of the array, make jumps, and reduce the jump length as we get closer to the target element.

Reviewer: Maybe mention you use powers of 2, like in binary jumping?

<JavaSection>

```java
public static int lastTrue(int lo, int hi, Predicate<Integer> f) {
	for (int dif = (hi-(--lo)); dif>0; dif /= 2)
		while (lo+dif <= hi && f.test(lo+dif)) lo += dif; //f is the function
	return lo;
}
```

</JavaSection>

<PySection>

<Warning>

> It is NOT recommended to use tenary expression in the python implementation of binary search because python tenary expression does not accept statements.

Reviewer: ```suggestion
It is NOT recommended to use ternary expressions in the Python implementation of binary search because Python ternary expressions do not accept statements.
```

<JavaSection>

```java
public static int lastTrue(int lo, int hi, Predicate<Integer> f) {

> 	for (int dif = (hi-(--lo)); dif>0; dif /= 2)

Reviewer 2: weird spacing in for loop, try 'dif > 0' with spaces instead of dif>0?

		v.push_back(t);
	}
	sort(v.begin(), v.end());

> 	cout << lastTrue(1,2e9);

Reviewer: ```suggestion
	cout << lastTrue(1, 2e9);
```
Keep spacing consistent

```

</CPPSection>

<JavaSection>

```java

> public static int firstTrue(int lo, int hi, Predicate<Integer> f) {

Reviewer 3: inconsistency

</CPPSection>

<JavaSection>

```java
public static void main(String[] args) {
	System.out.println(firstTrue(-10,-10,(x) -> false));
	// -8, should be -9
	System.out.println(firstTrue(-10,-10,(x) -> true));
	// infinite loop
}
```

</JavaSection>

<PySection>

```python
print(firstTrue(-10,-10,lambda x:False))
# -8, should be -9
print(firstTrue(-10,-10,lambda x:True))
# infinite loop
```

</PySection>

</LanguageSection>

This is because dividing an odd negative integer by two will round it up instead of down.

> We can fix this with:

Reviewer: No need for a new paragraph here

<JavaSection>

```java

> public static int firstTrue(Predicate<Integer> f, int lo, int hi) {

Reviewer 3: inconsistency

	for (hi ++; lo < hi; ) {
		int mid = lo+(hi-lo)/2;
		if(f.test(mid)) hi = mid; 
		else lo = mid+1;
	}
	return lo;
}
```

</JavaSection>

<PySection>

```python

> def firstTrue(lo, hi):

Reviewer 3: missing `f`?

</LanguageSection>

### Example - Counting Haybales

<FocusProblem problem={problems.count} />

> As each of the points are in the range $0 \ldots 1,000,000,000$, storing locations of haybales in a boolean array and then taking prefix sums of that would take too much time and memory.

Reviewer 3: ```suggestion
As each of the points are in the range $0 \ldots 1\,000\,000\,000$, storing locations of haybales in a boolean array and then taking prefix sums of that would take too much time and memory.
```

that's what cf uses and I think it looks nicer.
